### PR TITLE
tuyapi: Use Session object for every request made.

### LINKED ISSUE #36: DNS flood requests might be causing some issues on HA
Hello!

I'm posting this info here, since it seems to be an issue with the tuya api and how the requests are made to tuya servers.

After some versions, some people are starting to see an increased number of DNS requests via IPV6 (as reported https://github.com/home-assistant/core/issues/36713 and https://github.com/home-assistant/core/issues/26855), and I believe this is causing some issues that have been reported in HA (https://github.com/home-assistant/core/issues/36744) which makes the devices unusable because of a max retries request error.

```python
Jun 15 01:43:55 raspberrypi hass[4833]: urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='px1.tuyaus.com', port=443): Max retries exceeded with url: /homeassistant/skill (
Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0xa0314f70>: Failed to establish a new connection: [Errno 110] Connection timed out'))                   
Jun 15 01:43:55 raspberrypi hass[4833]: During handling of the above exception, another exception occurred:                                                                          
Jun 15 01:43:55 raspberrypi hass[4833]: Traceback (most recent call last):                                                                                                           
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 279, in async_update_ha_state                  
Jun 15 01:43:55 raspberrypi hass[4833]:     await self.async_device_update()                                                                                                         
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 472, in async_device_update                    
Jun 15 01:43:55 raspberrypi hass[4833]:     await self.hass.async_add_executor_job(self.update)                                                                                      
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run                                                                    
Jun 15 01:43:55 raspberrypi hass[4833]:     result = self.fn(*self.args, **self.kwargs)                                                                                              
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/components/tuya/__init__.py", line 254, in update                       
Jun 15 01:43:55 raspberrypi hass[4833]:     self._tuya.update()                                                                                                                      
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/tuyaha/devices/switch.py", line 23, in update                                         
Jun 15 01:43:55 raspberrypi hass[4833]:     devices = self.api.discovery()                                                                                                           
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/tuyaha/tuyaapi.py", line 115, in discovery                                            
Jun 15 01:43:55 raspberrypi hass[4833]:     response = self._request("Discovery", "discovery")                                                                                       
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/tuyaha/tuyaapi.py", line 161, in _request                                             
Jun 15 01:43:55 raspberrypi hass[4833]:     (TUYACLOUDURL + "/homeassistant/skill").format(SESSION.region), json=data                                                                
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/requests/api.py", line 119, in post                                                  
Jun 15 01:43:55 raspberrypi hass[4833]:     return request('post', url, data=data, json=json, **kwargs)                                                                             
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/requests/api.py", line 61, in request 
Jun 15 01:43:55 raspberrypi hass[4833]:     return session.request(method=method, url=url, **kwargs)
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/requests/sessions.py", line 530, in request
Jun 15 01:43:55 raspberrypi hass[4833]:     resp = self.send(prep, **send_kwargs)
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/requests/sessions.py", line 643, in send
Jun 15 01:43:55 raspberrypi hass[4833]:     r = adapter.send(request, **kwargs)
Jun 15 01:43:55 raspberrypi hass[4833]:   File "/srv/homeassistant/lib/python3.7/site-packages/requests/adapters.py", line 516, in send
Jun 15 01:43:55 raspberrypi hass[4833]:     raise ConnectionError(e, request=request)
Jun 15 01:43:55 raspberrypi hass[4833]: requests.exceptions.ConnectionError: HTTPSConnectionPool(host='px1.tuyaus.com', port=443): Max retries exceeded with url: /homeassistant/skil
l (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0xa0314f70>: Failed to establish a new connection: [Errno 110] Connection timed out'))
```

So, after a suggestion made here: https://github.com/home-assistant/core/issues/26855#issuecomment-609424984 I did the following monkey patch in `/srv/homeassistant/lib/python3.7/site-packages/tuyaha/tuyaapi.py`:

```python
import socket                                                                                                                                                                        
import requests.packages.urllib3.util.connection as urllib3_cn                                                                                                                       
                                                                                                                                                                                     
def allowed_gai_family():                                                                                                                                                            
    family = socket.AF_INET                                                                                                                                                          
    return family                                                                                                                                                                    
                                                                                                                                                                                     
urllib3_cn.allowed_gai_family = allowed_gai_family 
```

right after `_LOGGER`. It seems to be working, the DNS flood on the IPV6 land stopped and now  I can control most of my tuya devices without problems on HA.

I'm not sure where should I start taking a look for a proper fix, but I'm open to suggestions. Also, if the monkey patch is accepted as a solution, I can open a PR without problem.

### PR DESCRIPTION
**Pull request by estebanz01**
This change implements the Session object, allowing us to reuse TCP connections via requests pool.

First step in order to fully fix #36.

### CODE REVIEW

**Comment by estebanz01** [comment]
@PaulAnnekov I'm a ruby dev, so let me know if I'm missing something related to python styling.

Also, I was wondering if we can start doing the push approach in this PR or just open a new one.

**Comment by estebanz01** [comment]
This is what I wrote for testing, so now it maintains the final state and only one A query is done. Since I only have a switches, I do not have a real way to test the other devices, hopefully a smart lock will arrive soon 🤞 .

```python
from tuyaha import TuyaApi
import pprint
import time

USERNAME = "<value>"
PASSWORD = "<value>"
REGION = "<value>" # cn, eu, us
COUNTRY_CODE = "<value>" # Your account country code, e.g., 1 for USA or 86 for China
BIZ_TYPE = "<value>" # tuya, smart_life, jinvoo_smart

api = TuyaApi()

pp = pprint.PrettyPrinter(indent=4)

res = api.init(USERNAME, PASSWORD, COUNTRY_CODE, BIZ_TYPE)
res_1 = api.discover_devices()

pp.pprint(res)
pp.pprint(res_1)

print("turning on")
res[0].turn_on()
pp.pprint(res[0].state())
time.sleep(5)

print("turning off")
res[0].turn_off()
pp.pprint(res[0].state())
time.sleep(5)

print("turning on")
res[0].turn_on()
pp.pprint(res[0].state())
```

**Comment by djtimca** [comment]
I manually patched these files into my HA docker instance and restarted the container. Everything works much better now. 

Approximate delay when I update a switch from outside HA is ~30 seconds, but for HA controlled state changes it triggers instantly and keeps its state exactly as needed. 

One bug in light.py though:

In turn_off you missed the quotes around false so it returns an error in the UI (but does update after a while). I updated that manually and it works.

**Comment by estebanz01** [comment]
> I manually patched these files into my HA docker instance and restarted the container. Everything works much better now.
> 
> Approximate delay when I update a switch from outside HA is ~30 seconds, but for HA controlled state changes it triggers instantly and keeps its state exactly as needed.
> 
> One bug in light.py though:
> 
> In turn_off you missed the quotes around false so it returns an error in the UI (but does update after a while). I updated that manually and it works.

@djtimca thanks for testing this! 😁 I pushed a fix for the missing quotes.

**Comment by djtimca** [comment]
What is still required to merge / push the update so the HA Tuya component can be updated. Right now I have to manually update after every HA Update. 

@PaulAnnekov

**Comment by PaulAnnekov** [comment]
@djtimca I will try to take a look at this PR today.

File: tuyaha/tuyaapi.py [hunk 1]
```diff
@@ -14,6 +14,7 @@
 REFRESHTIME = 60 * 60 * 12
 
 _LOGGER = logging.getLogger(__name__)
+REQUEST_SESSION = requests.Session()
```
Comment (PaulAnnekov) [inline tuyaha/tuyaapi.py:17]: Please, move `REQUEST_SESSION` inside TuyaAPI.init method, like this:
```
self.requestSession = requests.Session()
```

do not pay attention to global `SESSION` variable. It's an old code from previous developers that needs refactoring.

Comment (estebanz01) [reply tuyaha/tuyaapi.py]: Fixed on ef4c333

**Comment by PaulAnnekov** [comment]
Thanks, good job!

**Pull request merged**